### PR TITLE
`aws-sam` launch configuration provider can provide default launch configurations for templated resources

### PR DESCRIPTION
**Pull request by bryceitoc9**
## Description
Added a basic implementation of `provideDebugConfigurations` to the `aws-sam` DebugConfigurationProvider. This does the following:
* If provided with a WorkspaceFolder, checks the CloudFormationTemplateRegistry for all templates within that folder
  * If the templates contain SAM functions, creates a launch config for each one with the following information:
```
    {
        type: AWS_SAM_DEBUG_TYPE,
        request: DIRECT_INVOKE_TYPE,
        name: resourceName,
        invokeTarget: {
            target: TEMPLATE_TARGET_TYPE,
            samTemplatePath: templatePath,
            samTemplateResource: resourceName
        }
    }
```
  * Returns the above in an array
* If no WorkspaceFolder is provided, returns `undefined`

FYI: please wait until https://github.com/aws/aws-toolkit-vscode/pull/957 merges; I'll pull in some changes from that.

## Motivation and Context
Completes `DebugConfigurationProvider` interface implementation

## Testing
Tests pass successfully

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the changelog using the script `npm run newChange`

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/sam/debugger/awsSamDebugger.ts [hunk 1]
```diff
@@ -40,3 +64,20 @@ export class AwsSamDebugConfigurationProvider implements vscode.DebugConfigurati
         return undefined
     }
 }
+
+function createSamDebugConfigurationFromTemplate(
+    resourceName: string,
+    templatePath: string,
+    resource: CloudFormation.Resource
```
Comment (awschristou) [inline src/shared/sam/debugger/awsSamDebugger.ts]: resource appears to be unused

File: src/shared/sam/debugger/awsSamDebugger.ts [hunk 2]
```diff
@@ -4,20 +4,44 @@
...
         token?: vscode.CancellationToken
     ): Promise<AwsSamDebuggerConfiguration[] | undefined> {
-        return undefined
+        if (folder) {
+            const debugConfigurations: AwsSamDebuggerConfiguration[] = []
+            const folderPath = folder.uri.fsPath
+            const templates = this.cftRegistry.registeredTemplates
+
+            for (const templateDatum of templates) {
+                if (templateDatum.path.startsWith(folderPath) && templateDatum.template.Resources) {
```
Comment (awschristou) [inline src/shared/sam/debugger/awsSamDebugger.ts]: can you double check that startswith isn't making a false positive match on a scenario where
workspace : `c:\foo`
template path : `c:\foobar\template.yaml`

Comment (bryceitoc9) [reply src/shared/sam/debugger/awsSamDebugger.ts]: Hmm, will do.

Comment (bryceitoc9) [reply src/shared/sam/debugger/awsSamDebugger.ts]: Done and has a test now.

File: src/shared/sam/debugger/awsSamDebugger.ts [hunk 3]
```diff
@@ -34,12 +35,19 @@ export class AwsSamDebugConfigurationProvider implements vscode.DebugConfigurati
             const templates = this.cftRegistry.registeredTemplates
 
             for (const templateDatum of templates) {
-                if (templateDatum.path.startsWith(folderPath) && templateDatum.template.Resources) {
+                const folderPathPieces = folderPath.split(sep)
```
Comment (awschristou) [inline src/shared/sam/debugger/awsSamDebugger.ts]: Move all the logic to support testing if a file is within a folder into a utility method
Motivations:
- reusable (and testable)
- easier to understand the logic that this function is performing

Comment (bryceitoc9) [reply src/shared/sam/debugger/awsSamDebugger.ts]: Done, moved to filesystemUtils

**Pull request merged**